lib/moab/storage_object_validator.rb

```diff
@@ -0,0 +1,181 @@
+    end
+
+    def validation_errors
+      errors = []
+      errors.concat check_correctly_named_version_dirs
+      errors.concat check_sequential_version_dirs if errors.empty?
+      errors.concat check_correctly_formed_moabs if errors.empty?
+      errors
+    end
+
+    # TODO: Figure out which methods should be public
+
+    private
+
+    def version_directories
+      @vdirs ||= sub_dirs(storage_obj_path)
+    end
+
+    def check_correctly_named_version_dirs
+      errors = []
+      version_directories.each do |version_dir|
+        if version_dir =~ /^[v]\d{4}/
+          nil
+        else
+          errors << result_hash(VERSION_DIR_BAD_FORMAT)
+        end
+      end
+      errors
+    end
+
+    # This method will be called only if the version directories are correctly named
+    def check_sequential_version_dirs
+      errors = []
+      version_directories.each_with_index do |dir_name, index|
+        expected_vers_num = index + 1 # version numbering starts at 1, array indexing starts at 0
+        if dir_name[1..-1].to_i != expected_vers_num
+          errors << result_hash(VERSIONS_NOT_IN_ORDER, version_directories)
+          break
+        end
+      end
+
+      errors
+    end
+
+    def check_correctly_formed_moabs
+      errors = []
+      version_directories.each do |version_dir|
+        version_path = "#{storage_obj_path}/#{version_dir}"
+        version_sub_dirs = sub_dirs(version_path)
+        before_result_size = errors.size
+        errors.concat check_sub_dirs(version_sub_dirs, version_dir, EXPECTED_VERSION_SUB_DIRS)
+        after_result_size = errors.size
+
+        if before_result_size == after_result_size
+          data_dir_path = "#{version_path}/#{EXPECTED_VERSION_SUB_DIRS[0]}"
+          data_sub_dirs = sub_dirs(data_dir_path)
+          errors.concat check_sub_dirs(data_sub_dirs, version_dir, EXPECTED_DATA_SUB_DIRS)
+          errors.concat check_required_manifest_files(version_path, version_dir)
+        end
+      end
+
+      errors
+    end
+
+    def check_sub_dirs(sub_dirs, version, required_sub_dirs)
+      errors = []
+      sub_dir_count = sub_dirs.size
+      if sub_dir_count == required_sub_dirs.size
+        errors.concat expected_dirs(sub_dirs, version, required_sub_dirs)
+      elsif sub_dir_count > required_sub_dirs.size
+        errors.concat found_unexpected(sub_dirs, version, required_sub_dirs)
+      elsif sub_dir_count < required_sub_dirs.size
+        errors.concat missing_dir(sub_dirs, version, required_sub_dirs)
+      end
+      errors
+    end
+
+    # This method removes the implicit '.' and '..' directories.
+    # Returns an array of strings.
+    def directory_entries(path)
+      @directory_entries_hash[path] ||=
+        begin
+          dirs = []
+          (Dir.entries(path).sort - IMPLICIT_DIRS).each do |child|
+            dirs << child
+          end
+          dirs
+        end
+    end
+
+    def sub_dirs(path)
+      directory_entries(path)
+    end
```

Comment:
I kinda left this like this just because I didn't incorporate the ``` [ ] content and metadata directories each contain only files ```.  
```
def directory_entries(path)
      @directory_entries_hash[path] ||=
        begin
          ret_val = { files: [], dirs: [] }
          (Dir.entries(path).sort - IMPLICIT_DIRS).each do |child|
            if FileTest.file? child
              ret_val[:files] << child
            else
              ret_val[:dirs] << child
            end
          end
          ret_val
        end
    end

    def sub_dirs(path)
      directory_entries(path)[:dirs]
    end

    def files_in_dir(path)
      directory_entries(path)[:files]
    end
```


 (you refactored that part, but i removed it for this PR. I created another ticket for that, so i could get this merged and a new gem release that was demoable. Also, the ``` File.file? ``` part of that code wasn't working properly so I just decided to work on that in another ticket ... )